adam/learner/integrated_learner.py

```diff
@@ -0,0 +1,348 @@
+
+
+class LanguageLearnerNew:
+    def observe(
+        self,
+        learning_example: LearningExample[
+            DevelopmentalPrimitivePerceptionFrame, LinguisticDescription
+        ],
+    ) -> None:
+        pass
+
+
+@attrs
+class IntegratedTemplateLearner(
+    LanguageLearner[
+        DevelopmentalPrimitivePerceptionFrame, TokenSequenceLinguisticDescription
+    ]
+):
+    """
+    A `LanguageLearner` which uses template-based syntax to learn objects, attributes, relations,
+    and actions all at once.
+    """
+
+    object_learner: TemplateLearner = attrib(validator=instance_of(TemplateLearner))
+    attribute_learner: Optional[TemplateLearner] = attrib(
+        validator=optional(instance_of(TemplateLearner)), default=None
+    )
+
+    relation_learner: Optional[TemplateLearner] = attrib(
+        validator=optional(instance_of(TemplateLearner)), default=None
+    )
+
+    action_learner: Optional[TemplateLearner] = attrib(
+        validator=optional(instance_of(TemplateLearner)), default=None
+    )
+
+    _max_attributes_per_word: int = attrib(validator=instance_of(int), default=3)
+
+    _observation_num: int = attrib(init=False, default=0)
+
+    def observe(
+        self,
+        learning_example: LearningExample[
+            DevelopmentalPrimitivePerceptionFrame, LinguisticDescription
+        ],
+    ) -> None:
+        logging.info(
+            "Observation %s: %s",
+            self._observation_num,
+            learning_example.linguistic_description.as_token_string(),
+        )
+
+        self._observation_num += 1
+
+        # We need to track the alignment between perceived objects
+        # and portions of the input language, so internally we operate over
+        # LanguageAlignedPerceptions.
+        current_learner_state = LanguagePerceptionSemanticAlignment(
+            language_concept_alignment=LanguageConceptAlignment.create_unaligned(
+                language=learning_example.linguistic_description
+            ),
+            perception_semantic_alignment=PerceptionSemanticAlignment(
+                perception_graph=self._extract_perception_graph(
+                    learning_example.perception
+                ),
+                semantic_nodes=[],
+            ),
+        )
+
+        # We iteratively let each "layer" of semantic analysis attempt
+        # to learn from the perception,
+        # and then to annotate the perception with any semantic alignments it knows.
+        for sub_learner in [
+            self.object_learner,
+            self.attribute_learner,
+            self.relation_learner,
+        ]:
+            if sub_learner:
+                # Currently we do not attempt to learn static things from dynamic situations
+                # because the static learners do not know how to deal with the temporal
+                # perception graph edge wrappers.
+                # @Reviewers: do not let me merge without making an issue for this.
```

Comment:
🗒️

Reply:
@spigo900 : Please create and reference the issue.

Reply:
@gabbard Done. It's issue #792.